caffe2/serialize/inline_container.h

```diff
@@ -119,135 +134,202 @@ class PyTorchStreamReader {
     auto ptr = malloc(size);
     at::DataPtr retval(ptr, ptr, free, at::kCPU);
 
-    in.read((char*)ptr, size);
-    cursor += size;
+    in_.read((char*)ptr, size);
+    cursor_ += size;
     seekToNextAlignmentBoundary();
     return std::tuple<at::DataPtr, size_t>(std::move(retval), size);
   }
+
+  // return dataptr, key, size
+  std::tuple<at::DataPtr, int64_t, int64_t> getNextRecord() {
+    int64_t key = cursor_;
+    if (!hasNextRecord()) {
+      throw std::runtime_error("No more record, but hasNextRecord is called.");
+    }
+    auto tag = read64BitIntegerLittleEndian();
+    if (tag != RecordTags::STORAGE) {
+      throw std::runtime_error(
+          "Attempted to read a record of non-storage type");
+    }
+    auto size = read64BitIntegerLittleEndian();
+    seekToNextAlignmentBoundary();
+    auto ptr = malloc(size);
+    at::DataPtr retval(ptr, ptr, free, at::kCPU);
+
+    in_.read((char*)ptr, size);
+    cursor_ += size;
+    seekToNextAlignmentBoundary();
+    return std::tuple<at::DataPtr, int64_t, int64_t>(
+        std::move(retval), key, size);
+  }
+
+  bool hasNextRecord() {
+    if (cursor_ + kFieldAlignment * 2 < file_size_) {
+      return true;
+    }
+    return false;
+  }
+
+  bool close() {
+    if (closed_) {
+      return false;
+    }
+    closed_ = true;
+    return true;
+  }
+
+  bool closed() {
+    return closed_;
+  }
+
   ~PyTorchStreamReader() {
   }
+
  private:
-  std::istream& in;
-  size_t cursor = 0;
-  size_t file_size;
-  size_t last_record_offset;
+  std::istream& in_;
```

Comment:
I'd suggest using a pointer instead of a reference for member variables. Reference is not a type while may lead to some potential issues.